Restrict api versions

irbekrm opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

This PR changes the cert-manager webhook configuration so that the Kubernetes API server will now convert legacy versions of cert-manager custom resources to v1 before sending them to the cert-manager admission webhooks.

This will help ensure that all users have a working conversion webhook and help ensure that all legacy versions of cert-manager CRs are upgraded to v1 before we drop support for the legacy versions of the API in cert-manager v1.6 and to avoid race conditions during upgrade to v1.6.

Note:
Initially I also  removed the v1beta1 versions from accepted `AdmissionReview` versions for mutating and validating webhook and accepted `ConversionReview` version for conversion webhook as part of this PR (as a cleanup), but it no longer seems like a good idea because it is still possible to serve the v1beta1 AdmissionReview API only in older versions of Kubernetes.

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
Pre-v1 cert-manager resource requests now must be converted to v1 in order to be validated/mutated by admission webhooks. (Default cert-manager validating and mutating webhook configurations ensure the resource requests are being converted)
```

Fixes #4113 
/kind cleanup

irbekrm commented:
/test pull-cert-manager-e2e-v1-16

wallrj reviewed (approved):
@irbekrm 

I installed cert-manager from this branch and tried applying some v1alpha2 Issuer and Certificate.
That worked.
And just to be sure that the v1alpha2 resources were still being sent for validation (despite the more specific `apiVersions` in this branch, I tried removing the dnsNames. 
The validation webhook denied the request. 

I read https://github.com/jetstack/cert-manager/pull/4021#pullrequestreview-664320743 and then went looking at:

```
kubectl explain mutatingwebhookconfiguration.webhooks.matchPolicy
KIND:     MutatingWebhookConfiguration
VERSION:  admissionregistration.k8s.io/v1

FIELD:    matchPolicy <string>

DESCRIPTION:
     matchPolicy defines how the "rules" list is used to match incoming
     requests. Allowed values are "Exact" or "Equivalent".

     - Exact: match a request only if it exactly matches a specified rule. For
     example, if deployments can be modified via apps/v1, apps/v1beta1, and
     extensions/v1beta1, but "rules" only included `apiGroups:["apps"],
     apiVersions:["v1"], resources: ["deployments"]`, a request to apps/v1beta1
     or extensions/v1beta1 would not be sent to the webhook.

     - Equivalent: match a request if modifies a resource listed in rules, even
     via another API group or version. For example, if deployments can be
     modified via apps/v1, apps/v1beta1, and extensions/v1beta1, and "rules"
     only included `apiGroups:["apps"], apiVersions:["v1"], resources:
     ["deployments"]`, a request to apps/v1beta1 or extensions/v1beta1 would be
     converted to apps/v1 and sent to the webhook.

     Defaults to "Equivalent"


```
And I see that we don't include that field so we get the the default value: `Equivalent`. 

So does that mean that:
* previously my `v1alpha2` resources would have been sent direct to the webhook/validate endpoint, without pre-conversion.
* And now, the resources are first sent to `webhook/convert` and then the `v1` resource is sent to `/validate`.
* So this change will add an extra round trip.

Will it also result in the `/convert` endpoint being called twice for each CREATE / UPDATE?

Out of interest, I edited the `validatingwebhookconfiguration.matchPolicy: Exact` and noticed that my `v1alpha2` Issuer and Certificate were no longer validated.
So if some user had patched our webhook configuration to use Exact (and I have no idea why they'd do that) then this change might mean that they are no longer validating cert-manager CRs.....so maybe it'd be worth adding a note to the website release notes and upgrade notes, explaining this....or maybe it'd just cause confusion.

One nit is that the description of this PR says:
> This PR restricts the mutating and validating webhooks to only accept v1 cert-manager.io and acme.cert-manager.io resources.

If I understand correctly it would be more accurate to say:
> This PR changes the cert-manager webhook *configuration* so that the Kubernetes API server will now convert legacy versions of cert-manager custom resources  to `v1` before sending them to the cert-manager validating webhook. 
>
> This will help ensure that all users have a working conversion webhook and help ensure that all legacy versions of cert-manager CRs are upgraded to `v1` before we drop support for the legacy versions of the API in cert-manager v1.6

LGTM, but I'll add a hold so that you have a chance to read my comments before you merge this.

/lgtm
/hold

wallrj commented:
I forgot to mention that I went looking at ClusterAPI to see whether other projects also specify a single API version in their webhooks. They do:

https://github.com/kubernetes-sigs/cluster-api/blob/a5e1b5fe003a29ec09a433cc5f6118e76993caa2/config/webhook/manifests.yaml#L22-L23

wallrj reviewed (approved):
/lgtm

irbekrm commented:
Thanks for the review and for digging into this @wallrj !

> Will it also result in the /convert endpoint being called twice for each CREATE / UPDATE?

It does look like it gets called for both validating and mutating webhook- I thought that perhaps the validating webhook would simply be sent the converted version. But I also think that users should look at converting their legacy resources in `v1.5`, so maybe it's not an issue.

> So if some user had patched our webhook configuration to use Exact (and I have no idea why they'd do that) then this change might mean that they are no longer validating cert-manager CRs.....so maybe it'd be worth adding a note to the website release notes and upgrade notes, explaining this....or maybe it'd just cause confusion.

I tried to add a release note that is not too alarming, but could potentially catch the attention of anyone doing that.
I imagine that this change might be visible to users getting webhook errors since now it is the conversion endpoint that gets called first.

> previously my v1alpha2 resources would have been sent direct to the webhook/validate endpoint, without pre-conversion

Yes, I think it would have been converted to the internal type inside the webhook, mutated/validated and converted back to `v1alpha2` ? See [this comment](https://github.com/jetstack/cert-manager/blob/master/pkg/internal/api/mutation/registry.go#L107-L109). 

I have also updated the PR description.

irbekrm commented:
/hold cancel
